File: R/soil_openlandmap.R
# Soil source: OpenLandMap / OpenGeoHub global soil COGs (live remote sampling).
# Samples cloud-optimized GeoTIFFs over HTTP via the OpenLandMap STAC catalog
# (URLs resolved at runtime). Texture from the 120 m global_soil_props product;
# bulk density + organic carbon from the 250 m COGs at the nearest point-depth.

OLM_STAC_BASE <- "https://s3.eu-central-1.wasabisys.com/stac/openlandmap/"
OLM_COLLECTIONS <- c(
  clay = "clay.tot_iso.11277.2020.wpct",
  sand = "sand.tot_iso.11277.2020.wpct",
  silt = "silt.tot_iso.11277.2020.wpct",
  bdod = "bulkdens.fineearth_usda.4a1h",
  oc   = "organic.carbon_usda.6a1c"
)
OLM_LAYERS <- data.frame(bottom = c(30, 60, 100), center = c(15, 45, 80))

.olm_depth_mid <- function(key) {
  m <- regmatches(key, regexec("b([0-9]+)cm\\.\\.([0-9]+)cm", key))[[1]]
  if (length(m) == 3) return((as.numeric(m[2]) + as.numeric(m[3])) / 2)
  m <- regmatches(key, regexec("b([0-9]+)cm", key))[[1]]
  if (length(m) == 2) return(as.numeric(m[2]))
  NA_real_
}

.olm_resolve_assets <- function(collection) {
  col <- jsonlite::fromJSON(paste0(OLM_STAC_BASE, collection, "/collection.json"), simplifyVector = FALSE)
  item_link <- Filter(function(l) identical(l$rel, "item"), col$links)[[1]]$href
  item_url <- paste0(OLM_STAC_BASE, collection, "/", sub("^\\./", "", item_link))
  assets <- jsonlite::fromJSON(item_url, simplifyVector = FALSE)$assets
  best <- list()
  for (k in names(assets)) {
    href <- assets[[k]]$href
    if (grepl("_m_", k) && grepl("\\.tif$", href) && !grepl("preview", k)) {
      mid <- .olm_depth_mid(k)
      if (!is.na(mid)) {
        fine <- grepl("120m", k)
        key <- as.character(mid)
        if (is.null(best[[key]]) || (fine && !best[[key]]$fine))
          best[[key]] <- list(mid = mid, href = href, fine = fine)
      }
    }
  }
  do.call(rbind, lapply(best, function(b) data.frame(mid = b$mid, href = b$href, stringsAsFactors = FALSE)))
}

.olm_nearest <- function(assets, target) assets$href[which.min(abs(assets$mid - target))]

.olm_sample <- function(url, pts_vect) {
  r <- tryCatch(terra::rast(paste0("/vsicurl/", url)), error = function(e) NULL)
  if (is.null(r)) return(rep(NA_real_, nrow(pts_vect)))
  v <- terra::extract(r, pts_vect, ID = FALSE)[, 1]
  nd <- terra::NAflag(r)
  if (!is.na(nd)) v[v == nd] <- NA_real_
  as.numeric(v)
}

process_soils_openlandmap <- function(grid_points, output_csv_path, output_sol_dir,
                                      id_col = "ID", lat_col = "LAT", long_col = "LONG") {
  message("--- Starting OpenLandMap COG Extraction (live) ---")
  pts <- sf::st_transform(grid_points, 4326)
  ids <- as.character(sf::st_drop_geometry(pts)[[id_col]])
  xy <- sf::st_coordinates(pts); lats <- xy[, 2]; lons <- xy[, 1]
  pts_vect <- terra::vect(pts)
  assets <- lapply(OLM_COLLECTIONS, .olm_resolve_assets)

  rows <- list()
  for (j in seq_len(nrow(OLM_LAYERS))) {
    dctr <- OLM_LAYERS$center[j]; dbot <- OLM_LAYERS$bottom[j]
    clay <- .olm_sample(.olm_nearest(assets$clay, dctr), pts_vect)
    sand <- .olm_sample(.olm_nearest(assets$sand, dctr), pts_vect)
    silt <- .olm_sample(.olm_nearest(assets$silt, dctr), pts_vect)
    bd <- .olm_sample(.olm_nearest(assets$bdod, dctr), pts_vect)
    bd <- ifelse(bd > 10, bd / 100, bd)
    oc <- .olm_sample(.olm_nearest(assets$oc, dctr), pts_vect)
    soc_pct <- oc / 50  # x5 g/kg -> g/kg (/5) -> percent (/10)
    silt <- ifelse(!is.finite(silt) & is.finite(sand) & is.finite(clay),
                   pmax(0, 100 - sand - clay), silt)
    rows[[j]] <- data.frame(ID = ids, latitude = lats, longitude = lons,
                            depth_bottom = dbot, depth_center = dctr,
                            sand = sand, clay = clay, silt = silt,
                            bdod = bd, soc_pct = soc_pct, cfvo = 0)
  }
  df <- do.call(rbind, rows)
  df <- df[complete.cases(df[, c("sand", "clay", "bdod", "soc_pct")]), ]
  if (!nrow(df)) stop("No usable OpenLandMap data sampled. Check connectivity / coordinates.")
  df <- soil_add_physics(df)
  soil_write_mapping(ids, output_csv_path)
  soil_write_profiles(df, output_sol_dir, "OpenLandMap (OpenGeoHub)", "STAC COG sampling")
}

File: python/dssatutils/soil_openlandmap.py
# Soil source: OpenLandMap / OpenGeoHub global soil COGs (live remote sampling).
#
# Samples cloud-optimized GeoTIFFs over HTTP (no bulk download) via the OpenLandMap
# STAC catalog, so URLs are resolved at runtime and survive version bumps. Texture
# (clay/sand/silt) comes from the global_soil_props product (120 m, 0-30/30-60/
# 60-100 cm); bulk density and organic carbon from the 250 m property COGs at the
# point-depth nearest each layer centre. Water limits are derived via Saxton-Rawls.
#
# Catalog: https://s3.eu-central-1.wasabisys.com/stac/openlandmap/catalog.json

import re

import numpy as np
import pandas as pd
import requests

from .soil_raster_common import (
    add_physics_texture, coords, normalize_bd, write_mapping, write_profiles,
)

_STAC_BASE = "https://s3.eu-central-1.wasabisys.com/stac/openlandmap/"
# property -> STAC collection id
_COLLECTIONS = {
    "clay": "clay.tot_iso.11277.2020.wpct",
    "sand": "sand.tot_iso.11277.2020.wpct",
    "silt": "silt.tot_iso.11277.2020.wpct",
    "bdod": "bulkdens.fineearth_usda.4a1h",
    "oc":   "organic.carbon_usda.6a1c",
}
# DSSAT layers (depth_bottom_cm, depth_center_cm) matching the texture product.
_LAYERS = [(30, 15.0), (60, 45.0), (100, 80.0)]


def _depth_mid(key: str):
    """Parse a depth midpoint (cm) from a STAC asset key like 'b0cm..30cm' or 'b10cm'."""
    m = re.search(r"b(\d+)cm\.\.(\d+)cm", key)
    if m:
        return (int(m.group(1)) + int(m.group(2))) / 2.0
    m = re.search(r"b(\d+)cm", key)
    return float(m.group(1)) if m else None


def _resolve_assets(collection: str):
    """Return [(depth_mid_cm, cog_url)] for the mean COG assets of a collection."""
    col = requests.get(f"{_STAC_BASE}{collection}/collection.json", timeout=60).json()
    item_link = next(l for l in col["links"] if l.get("rel") == "item")
    item_url = requests.compat.urljoin(f"{_STAC_BASE}{collection}/", item_link["href"])
    assets = requests.get(item_url, timeout=60).json()["assets"]
    out = []
    for k, v in assets.items():
        href = v.get("href", "")
        # mean COGs only (key has '_m_'), prefer the finer 120 m where present.
        if "_m_" in k and href.endswith(".tif") and "preview" not in k:
            mid = _depth_mid(k)
            if mid is not None:
                out.append((mid, href, "120m" in k))
    # prefer 120 m assets when both resolutions exist for the same depth
    best = {}
    for mid, href, fine in out:
        if mid not in best or (fine and not best[mid][1]):
            best[mid] = (href, fine)
    return sorted((mid, href) for mid, (href, _) in best.items())


def _sample_cog(url: str, lats, lons):
    import rasterio
    from pyproj import Transformer
    out = np.full(len(lats), np.nan, dtype=float)
    with rasterio.open("/vsicurl/" + url) as src:
        dst = src.crs.to_string() if src.crs else "EPSG:4326"
        xs, ys = Transformer.from_crs("EPSG:4326", dst, always_xy=True).transform(lons, lats)
        nd = src.nodata
        for i, cell in enumerate(src.sample(zip(xs, ys))):
            v = float(cell[0])
            if nd is not None and v == float(nd):
                continue
            out[i] = v
    return out


def _nearest_asset(assets, target_mid):
    return min(assets, key=lambda a: abs(a[0] - target_mid))[1]


def process_soils_openlandmap(
    grid_points, output_csv_path: str, output_sol_dir: str,
    id_col: str = "ID", lat_col: str = "LAT", long_col: str = "LONG",
) -> None:
    """Build DSSAT .SOL files by sampling OpenLandMap COGs over HTTP (live).

    No local data required; resolves COG URLs from the OpenLandMap STAC catalog
    and samples each point. Bulk density is scaled to g/cm3 and organic carbon
    (stored as x5 g/kg) to percent before the Saxton-Rawls water-limit step.
    """
    ids, lats, lons = coords(grid_points, id_col, lat_col, long_col)
    print("--- Starting OpenLandMap COG Extraction (live) ---")
    assets = {p: _resolve_assets(c) for p, c in _COLLECTIONS.items()}

    rows = []
    for dbot, dctr in _LAYERS:
        # texture intervals align with layer; BD/OC use nearest point-depth.
        clay = _sample_cog(_nearest_asset(assets["clay"], dctr), lats, lons)
        sand = _sample_cog(_nearest_asset(assets["sand"], dctr), lats, lons)
        silt = _sample_cog(_nearest_asset(assets["silt"], dctr), lats, lons)
        bd = normalize_bd(_sample_cog(_nearest_asset(assets["bdod"], dctr), lats, lons))
        oc = _sample_cog(_nearest_asset(assets["oc"], dctr), lats, lons)
        soc_pct = oc / 50.0  # x5 g/kg -> g/kg (/5) -> percent (/10)
        for i, pid in enumerate(ids):
            sa, si, cl = sand[i], silt[i], clay[i]
            if np.isfinite(sa) and np.isfinite(cl) and not np.isfinite(si):
                si = max(0.0, 100.0 - sa - cl)
            rows.append({
                "ID": pid, "latitude": lats[i], "longitude": lons[i],
                "depth_bottom": dbot, "depth_center": dctr,
                "sand": sa, "clay": cl, "silt": si, "bdod": bd[i],
                "soc_pct": soc_pct[i], "cfvo": 0.0,
            })
    df = pd.DataFrame(rows).dropna(subset=["sand", "clay", "bdod", "soc_pct"])
    if df.empty:
        raise RuntimeError("No usable OpenLandMap data sampled. Check connectivity / coordinates.")
    df = add_physics_texture(df)
    write_mapping(ids, output_csv_path)
    write_profiles(df, output_sol_dir, "OpenLandMap (OpenGeoHub)", "STAC COG sampling")
